### PR TITLE
mgr/dashboard: RGW multisite sync flow

### PR DESCRIPTION
**Pull request by nmunet**
Fixes: https://tracker.ceph.com/issues/66915

Signed-off-by: Naman Munet <nmunet@redhat.com>

Sync Flow Demo - 

https://github.com/user-attachments/assets/be20ac03-c182-44c9-bb3d-1a9bd6825bc2

Directional Sync flow edit/delete is not possible yet as the core backend does not sends flow_id in response, which is necessary for the operations. Hence, kept disabled for now.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [x] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by nmunet** [comment]
There are still some refactoring left with some e2e, WIP.

**Comment by cloudbehl** [comment]
Thanks for PR. Mostly looks good! I have not tested it yet. 

Few questions:
1. Why do we have bucket selection first and flow Id second when we bucket is not even mandatory?
2. Why is bucket not a required parameter and what happens when value is not set?
3. Based on the bucket selected, can we select the zone automatically in which that bucket exists? 

Can we change `Flow Id` to `Name` and make it a first parameter?

**Comment by nmunet** [comment]
jenkins test dashboard

**Comment by nmunet** [comment]
jenkins test api

**Comment by nmunet** [comment]
jenkins test make check

**Comment by nmunet** [comment]
jenkins test make check arm64

**Comment by nmunet** [comment]
> Thanks for PR. Mostly looks good! I have not tested it yet.
> 
> Few questions:
> 
> 1. Why do we have bucket selection first and flow Id second when we bucket is not even mandatory?
> 2. Why is bucket not a required parameter and what happens when value is not set?
> 3. Based on the bucket selected, can we select the zone automatically in which that bucket exists?
> 
> Can we change `Flow Id` to `Name` and make it a first parameter?

1) I have addressed this issue in my latest commit
2) we can only set bucket value if the sync policy is created with a bucket name, else not. And if we do not provide a bucket name it creates the sync policy/flow/pipe on the zonegroup level.
3) If we are talking about symmetrical flow then the zones could be multisite, in case of directional maybe we can select zone in which bucket exists.
4) Changed 'flow id' to 'name'

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-sync-flow-modal/rgw-multisite-sync-flow-modal.component.html [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+          <div class="form-group row">
+            <label
+              class="cd-col-form-label required"
+              for="flow_id"
+              i18n>Name</label>
+            <div class="cd-col-form-input">
+              <input
+                class="form-control"
+                type="text"
+                placeholder="Flow Id..."
```
Comment (cloudbehl) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-sync-flow-modal/rgw-multisite-sync-flow-modal.component.html:22]:
```suggestion
                placeholder="Flow Name..."
```

File: src/pybind/mgr/dashboard/frontend/src/app/shared/services/task-message.service.ts [hunk 2]
```diff
@@ -334,6 +334,14 @@ export class TaskMessageService {
         }`;
       }
     ),
+    'rgw/multisite/sync-flow/delete': this.newTaskMessage(
+      this.commonOperations.delete,
+      (metadata) => {
+        return $localize`${
+          metadata.flow_ids.length > 1 ? 'selected Flow Ids' : `Flow Id '${metadata.flow_ids[0]}'`
```
Comment (cloudbehl) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/services/task-message.service.ts:341]:
```suggestion
          metadata.flow_ids.length > 1 ? 'selected Flow Names' : `Flow Name '${metadata.flow_ids[0]}'`
```

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-sync-policy-details/rgw-multisite-sync-policy-details.component.ts [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+      {
+        name: 'Source Zone',
+        prop: 'source_zone',
+        flexGrow: 1
+      },
+      {
+        name: 'Destination Zone',
+        prop: 'dest_zone',
+        flexGrow: 1
+      }
```
Comment (cloudbehl) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-sync-policy-details/rgw-multisite-sync-policy-details.component.ts:74]: Why are we not showing name of directional flow?
```suggestion
      {
        name: 'Name',
        prop: 'id',
        flexGrow: 1
      },
      {
        name: 'Source Zone',
        prop: 'source_zone',
        flexGrow: 1
      },
      {
        name: 'Destination Zone',
        prop: 'dest_zone',
        flexGrow: 1
      }
```

**Review by cloudbehl — APPROVED** [review]
LGTM

Some nits:

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-sync-policy-details/rgw-multisite-sync-policy-details.component.ts (hunk 3 above)
Comment (nmunet) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-sync-policy-details/rgw-multisite-sync-policy-details.component.ts:74]: we are not getting flow name from the API, there is already a bz opened for this issue.
That's why we are disabling the directional flow edit  delete functionality for now!

Comment (nmunet) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-sync-policy-details/rgw-multisite-sync-policy-details.component.ts:74]: mentioned this in my PR description already!

Comment (cloudbehl) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-sync-policy-details/rgw-multisite-sync-policy-details.component.ts:74]: Ah I missed it!

**Comment by nmunet** [comment]
Updated PR with e2es and review suggestions!

**Comment by nmunet** [comment]
jenkins test dashboard cephadm

**Comment by nmunet** [comment]
jenkins test dashboard

**Comment by nmunet** [comment]
jenkins test make check

**Comment by nmunet** [comment]
jenkins test make check arm64

**Comment by nmunet** [comment]
jenkins test dashboard cephadm

**Comment by nmunet** [comment]
jenkins test dashboard

**Comment by nmunet** [comment]
jenkins test api

**Comment by nmunet** [comment]
jenkins test make check

**Comment by nmunet** [comment]
jenkins test dashboard

**Comment by nmunet** [comment]
jenkins test dashboard

**Comment by nmunet** [comment]
jenkins test dashboard cephadm

**Comment by nmunet** [comment]
jenkins test dashboard cephadm

**Comment by nmunet** [comment]
jenkins test dashboard

**Comment by nmunet** [comment]
jenkins test make check arm64

**Comment by nmunet** [comment]
jenkins test dashboard

**Review by nizamial09 — APPROVED** [review]

**Pull request merged**